src/util/json_stream.cpp

```diff
@@ -0,0 +1,142 @@
+/*******************************************************************\
+
+Module:
+
+Author: Peter Schrammel
+
+\*******************************************************************/
+
+#include "json_stream.h"
+
+#include <ostream>
+
+/// Out a delimiter to the current open JSON stream.
+void json_streamt::output_delimiter()
+{
+  if(!first)
+    out << ',';
+  else
+    first = false;
+  out << '\n';
+  out << std::string((indent+1)*2, ' ');
+}
+
+/// Construct a new JSON array stream object.
+json_stream_arrayt::json_stream_arrayt(std::ostream &out, unsigned indent)
+  : json_streamt(out, indent)
+{
+  out << '[';
+}
+
+/// Output current JSON array and close stream.
+void json_stream_arrayt::output_current()
+{
+  if(!object.empty())
+  {
+    output_delimiter();
+    object[""].output_rec(out, indent+1);
+    object.clear();
+  }
+  if(current)
+  {
+    current->close();
+  }
+}
+
+/// Output a JSON object finalizing character for a JSON array stream.
+void json_stream_arrayt::output_finalizer()
+{
+  out << '\n' << std::string(indent*2, ' ');
+  out << ']';
+}
+
+/// Constructor for json_stream_objectt.
+/// \param out: The stream that is to be used to output the JSON object.
+/// \param indent: Current indentation level.
+json_stream_objectt::json_stream_objectt(std::ostream &out, unsigned indent)
+  : json_streamt(out, indent)
+{
+  out << '{';
+}
+
+/// Create a new JSON array stream.
+json_stream_arrayt &json_streamt::create_current_array()
+{
+  current =
+    std::unique_ptr<json_streamt>(new json_stream_arrayt(out, indent + 1));
+  return static_cast<json_stream_arrayt &>(*current);
+}
+
+/// Create a new JSON object stream.
+json_stream_objectt &json_streamt::create_current_object()
+{
+  current =
+    std::unique_ptr<json_streamt>(new json_stream_objectt(out, indent + 1));
+  return static_cast<json_stream_objectt &>(*current);
+}
```

Comment:
This code is really hard to read: lots of bits and pieces that don't seem to fit together in any obvious way.

Reply:
I agree. There is some doxygen work to be done to explain how things fit together.